File: I semestr/HomeWork8/Zadacha1 hash table/Zadacha1 hash/list.cpp
#include <iostream>
#include "list.h"

using namespace std;

struct list::ListElement
{
	ElementType value;
	int quantity;
	ListElement *next;
};

struct list::List 
{
	ListElement *head;
};

list::List *list::create()
{
	List *result = new List;
	result->head = nullptr;
	return result;
}

void list::insert(List *list, ElementType value, Position position)
{
	ListElement *newElement = new ListElement;
	newElement->value = value;
	newElement->quantity = 1;
	newElement->next = position->next;
	position->next = newElement;
}

void list::insertToHead(List *list, ElementType value)
{
	ListElement *newElement = new ListElement;
	newElement->value = value;
	newElement->quantity = 1;
	newElement->next = list->head;
	list->head = newElement;
}

void list::removeList(List *list)
{
	ListElement *position = list->head;
	while (position != nullptr)
	{
		ListElement *temp = position;
		position = position->next;
		delete temp->value;
		delete temp;
	}
	delete list;
}

void list::remove(List *list, Position position)
{
	if (position == nullptr) 
	{
		cout << "Error" << endl;
		return;
	}
	Position temp = position;
	if (position == list->head)
		list->head = list->head->next;
	else
	{
		position = list->head;
		while (position->next != temp)
			position = position->next;
		position->next = temp->next;
	}
	delete temp;
}

list::Position list::positionOfElement(List *list, ElementType value)
{
	Position position = list->head;
	while (position != nullptr)
	{
		if (strcmp(position->value, value) == 0)
			return position;
		else
		{
			position = position->next;
		}
	}
	return nullptr;
}

void list::print(List *list)
{
	Position position = list->head;
	while (position != nullptr)
	{
		cout << position->value << endl;
		cout << "Number of repetition = " << position->quantity << "\n" << endl;
		position = position->next;
	}
}

bool list::exist(List *list, ElementType value)
{
	ListElement *position = list->head;
	while (position != nullptr)
	{
		if (strcmp(position->value, value) == 0)
			return true;
		else
			position = position->next;
	}
	return false;
}


list::Position list::first(List *list)
{
	return list->head;
}

list::Position list::end(List *list)
{
	return nullptr;
}
list::Position list::next(List *list, Position position)
{
	return position->next;
}
list::ElementType list::valueOnPosition(List *list, Position position)
{
	return position->value;
}

void list::quantityOfElementPlusOne(List *list, ElementType element)
{
	Position position = list->head;
	while (position != nullptr && strcmp(position->value, element) != 0)
	{
		position = position->next;
	}
	if (position != nullptr)
		position->quantity++;
	return;
}

File: II semestr/HomeWork7/Zadacha1 Generic List and Stack/Zadacha1 Generic List and Stack/List.cs
namespace ListNamespace
{
    using System;

    public class List<T>
    {
        private class ListElement
        {
            public ListElement Next { get; set; }

            public T Value { get; set; }
        }

        private ListElement head;

        /// <summary>
        /// Insert value in the head of the list.
        /// </summary>
        /// <param name="value"></param>
        public void InsertToHead(T value)
        {
            var newElement = new ListElement()
            {
                Value = value,
                Next = head
            };
            head = newElement;
        }

        /// <summary>
        /// Return value of the position.
        /// If position bigger than number of elements in list then return -1.
        /// </summary>
        /// <param name="position"></param>
        /// <returns></returns>
        public T ValueOfPosition(int position)
        {
            var element = head;
            if (element == null)
            {
                return default(T);
            }
            for (int i = 0; i < position; i++)
            {
                if (element.Next == null)
                {
                    return default(T);
                }
                element = element.Next;
            }
            return element.Value;
        }

        /// <summary>
        /// Insert value to position in list.
        /// If list is empty then insert to head.
        /// If number of elements in list less than posision then insert to end of list.
        /// </summary>
        /// <param name="value"></param>
        /// <param name="position"></param>
        public void InsertToPosition(T value, int position)
        {
            var element = head;
            if (element == null || position == 0)
            {
                InsertToHead(value);
                return;
            }
            for (int i = 0; i < position - 1; i++)
            {
                if (element.Next == null)
                {
                    break;
                }
                element = element.Next;
            }
            var newElement = new ListElement()
            {
                Value = value,
                Next = element.Next
            };
            element.Next = newElement;
        }

        /// <summary>
        /// Remove element to position in list.
        /// If list is empty then this will be written to the console.
        /// If position bigger then number of elements of the list then this will be written.
        /// </summary>
        /// <param name="position"></param>
        public void RemoveElementToPosition(int position)
        {
            if (position == 0)
            {
                head = head.Next;
                return;
            }
            var element = head;
            if (element == null)
            {
                Console.WriteLine("List is empty!");
                return;
            }
            for (int i = 0; i < position - 1; i++)
            {
                if (element.Next == null)
                {
                    Console.WriteLine("There is not element in this position!");
                    return;
                }
                element = element.Next;
            }
            element.Next = element.Next.Next;
        }

        /// <summary>
        /// Print all list on console.
        /// </summary>
        public void Print()
        {
            var element = head;
            while (element != null)
            {
                Console.WriteLine("Element = {0}", element.Value);
                element = element.Next;
            }
        }

        /// <summary>
        /// Clean all list.
        /// </summary>
        public void RemoveList()
        {
            head = null;
        }

        /// <summary>
        /// Return 'true' if list is empty and 'false' if it is not empty.
        /// </summary>
        /// <returns></returns>
        public bool IsEmpty()
        {
            return (head == null);
        }

        /// <summary>
        /// Is value in the list?
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        public bool Exist(T value)
        {
            var element = head;
            while (element != null)
            {
                if (Equals(element.Value, value))
                {
                    return true;
                }
                element = element.Next;
            }
            return false;
        }
    }
}
